### PR TITLE
Resolves #13629 Add RegistryAuthHeader to manifest push

### LINKED ISSUE #13629: podman -r manifest push does not work unless I log into quay within the machine vm first
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

I've set up a podman machine, done a multiarch build, and tried to push it using the machine. It seems to work up until I run podman -r manifest push.

**Steps to reproduce the issue:**

1. podman machine init --cpus 2 -m 8192
2. podman machine start
3. podman machine ssh sudo rpm-ostree install qemu-user-static
4. podman machine ssh sudo systemctl reboot
5. git clone https://github.com/jwmatthews/case_watcher.git
6. cd case_watcher
7. podman -r build -f Dockerfile --manifest quay.io/jmontleon/case_watcher:latest --platform linux/amd64,linux/arm64 .
8. podman -r manifest push quay.io/jmontleon/case_watcher:latest docker://quay.io/jmontleon/case_watcher:latest

**Describe the results you received:**
```
$ podman -r --log-level=debug manifest push quay.io/jmontleon/case_watcher:latest docker://quay.io/jmontleon/case_watcher:latest
INFO[0000] podman filtering at log level debug          
DEBU[0000] Called push.PersistentPreRunE(podman -r --log-level=debug manifest push quay.io/jmontleon/case_watcher:latest docker://quay.io/jmontleon/case_watcher:latest) 
DEBU[0000] SSH Ident Key "/home/jason/.ssh/podman-machine-default" SHA256:S4G1wRA6yAbG7O/Tq15fzTnWekJmCdB0tu4BAz8Q5zQ ssh-ed25519 
DEBU[0000] Found SSH_AUTH_SOCK "/run/user/1000/ssh-agent.socket", ssh-agent signer(s) enabled 
DEBU[0000] SSH Agent Key SHA256:i13XTbtJgcZoJn2oXAYp7FJvBHqqJL9nPC3n1WdyEsA ssh-rsa 
DEBU[0000] SSH Agent Key SHA256:rJsMx7sPGxhckIxUh4mjhDBgQwAhr0QG6F9eBoDXUZo ssh-dss 
DEBU[0000] SSH Agent Key SHA256:DoYGRPSmnKydjI5kfbFmZDkgq9Df/V++1QUrmEdlwsg ecdsa-sha2-nistp521 
DEBU[0000] DoRequest Method: GET URI: http://d/v3.4.4/libpod/_ping 
DEBU[0000] SSH Ident Key "/home/jason/.ssh/podman-machine-default" SHA256:S4G1wRA6yAbG7O/Tq15fzTnWekJmCdB0tu4BAz8Q5zQ ssh-ed25519 
DEBU[0000] Found SSH_AUTH_SOCK "/run/user/1000/ssh-agent.socket", ssh-agent signer(s) enabled 
DEBU[0000] SSH Agent Key SHA256:i13XTbtJgcZoJn2oXAYp7FJvBHqqJL9nPC3n1WdyEsA ssh-rsa 
DEBU[0000] SSH Agent Key SHA256:rJsMx7sPGxhckIxUh4mjhDBgQwAhr0QG6F9eBoDXUZo ssh-dss 
DEBU[0000] SSH Agent Key SHA256:DoYGRPSmnKydjI5kfbFmZDkgq9Df/V++1QUrmEdlwsg ecdsa-sha2-nistp521 
DEBU[0000] DoRequest Method: GET URI: http://d/v3.4.4/libpod/_ping 
DEBU[0000] DoRequest Method: POST URI: http://d/v3.4.4/libpod/manifests/quay.io%2Fjmontleon%2Fcase_watcher:latest/push 
DEBU[0001] Called push.PersistentPostRunE(podman -r --log-level=debug manifest push quay.io/jmontleon/case_watcher:latest docker://quay.io/jmontleon/case_watcher:latest) 
```
This returns very fast and nothing is pushed.

**Describe the results you expected:**
The manifest actually gets pushed.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `podman version`:**

```
$ podman version
Version:      3.4.4
API Version:  3.4.4
Go Version:   go1.16.8
Built:        Wed Dec  8 16:45:07 2021
OS/Arch:      linux/amd64
```

**Output of `podman info --debug`:**

```
host:
  arch: amd64
  buildahVersion: 1.23.1
  cgroupControllers:
  - memory
  - pids
  cgroupManager: systemd
  cgroupVersion: v2
  conmon:
    package: conmon-2.1.0-2.fc35.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.1.0, commit: '
  cpus: 4
  distribution:
    distribution: fedora
    version: "35"
  eventLogger: journald
  hostname: pixelbook.montleon.intra
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
  kernel: 5.16.16-201.pixelbook.fc35.x86_64
  linkmode: dynamic
  logDriver: journald
  memFree: 271175680
  memTotal: 16659902464
  ociRuntime:
    name: crun
    package: crun-1.4.3-1.fc35.x86_64
    path: /usr/bin/crun
    version: |-
      crun version 1.4.3
      commit: 61c9600d1335127eba65632731e2d72bc3f0b9e8
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +CRIU +YAJL
  os: linux
  remoteSocket:
    path: /run/user/1000/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: true
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: true
  serviceIsRemote: false
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: slirp4netns-1.1.12-2.fc35.x86_64
    version: |-
      slirp4netns version 1.1.12
      commit: 7a104a101aa3278a2152351a082a6df71f57c9a3
      libslirp: 4.6.1
      SLIRP_CONFIG_VERSION_MAX: 3
      libseccomp: 2.5.3
  swapFree: 25768484864
  swapTotal: 25769795584
  uptime: 11h 54m 12.23s (Approximately 0.46 days)
plugins:
  log:
  - k8s-file
  - none
  - journald
  network:
  - bridge
  - macvlan
  volume:
  - local
registries:
  search:
  - registry.fedoraproject.org
  - registry.access.redhat.com
  - docker.io
  - quay.io
store:
  configFile: /home/jason/.config/containers/storage.conf
  containerStore:
    number: 0
    paused: 0
    running: 0
    stopped: 0
  graphDriverName: overlay
  graphOptions: {}
  graphRoot: /home/jason/.local/share/containers/storage
  graphStatus:
    Backing Filesystem: xfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
    Using metacopy: "false"
  imageStore:
    number: 0
  runRoot: /run/user/1000/containers
  volumePath: /home/jason/.local/share/containers/storage/volumes
version:
  APIVersion: 3.4.4
  Built: 1638999907
  BuiltTime: Wed Dec  8 16:45:07 2021
  GitCommit: ""
  GoVersion: go1.16.8
  OsArch: linux/amd64
  Version: 3.4.4
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
podman-3.4.4-1.fc35.x86_64
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshing Guide? (https://github.com/containers/podman/blob/main/troubleshooting.md)**

No. I reproduced this trying to help @jwmatthews experiencing the same problem on a Mac using 4.0.2.

**Additional environment details (AWS, VirtualBox, physical, etc.):**
If I do a normal build `podman -r build -f Dockerfile . -t ....` I can run `podman -r push ...` and it will push without first SSHing into the VM and running podman login.

If I do a multiarch build it seems like it will only work if I ssh into the vm log in to quay, exit, and then run `podman -r manifest push ...`

### PR DESCRIPTION
**Pull request by jmontleon**
Signed-off-by: Jason Montleon <jmontleo@redhat.com>

### CODE REVIEW

**Comment by rhatdan** [comment]
You are either going to need to add a test or add the 
[NO NEW TESTS NEEDED] to your PR so the test will pass.

**Comment by rhatdan** [comment]
@mtrmac @vrothberg PTAL

File: test/e2e/manifest_test.go [hunk 1]
```diff
@@ -272,6 +272,55 @@ var _ = Describe("Podman manifest", func() {
...
+		session.WaitWithDefaultTimeout()
+		Expect(session).Should(Exit(0))
+		session = podmanTest.Podman([]string{"pull", ALPINE})
+		session.WaitWithDefaultTimeout()
+		Expect(session).Should(Exit(0))
+
+		lock := GetPortLock("5000")
+		defer lock.Unlock()
+
+		session = podmanTest.Podman([]string{"run", "-d", "-p", "5000:5000", "--name", "registry", "-e", "REGISTRY_AUTH=htpasswd", "-e",
```
Comment (vrothberg) [inline test/e2e/manifest_test.go:289]: Note to @containers/podman-maintainers:

We should migrate the tests to be using `github.com/containers/podman/v4/hack/podman-registry-go`: https://github.com/containers/podman/blob/main/pkg/bindings/test/auth_test.go#L10

**Review by vrothberg — COMMENTED** [review]
Changes LGTM, thank you.

`manifest add` doesn't seem to work yet though.

@jwhonce PTAL

**Comment by vrothberg** [comment]
Could you also squash the two commits into one?

**Comment by jmontleon** [comment]
I squashed the commits and I think I fixed `add` as well. We'll see what the tests say.

**Comment by jmontleon** [comment]
@vrothberg @rhatdan The add commands are working for me locally, but I had to replace podman on the remote machine/vm because of the changes to the handler.

Is this accounted for in the tests and I've messed something up, or will there need to be an updated image for the tests somehow?

**Comment by jmontleon** [comment]
There's something else going on with the Handlers. It seems like the v4 create handler is going off for every post request. I'm not sure what to make of it; seems like something is very wrong.

**Comment by jwhonce** [comment]
@vrothberg Can the migration to hack/podman-registry-go be a follow-on PR?

**Comment by jmontleon** [comment]
I can have a look and try to adapt it for the PR. Hopefully not too hard now that it's working.

File: contrib/cirrus/setup_environment.sh [hunk 2]
```diff
@@ -36,6 +36,8 @@ do
     fi
 done
 
+cp hack/podman-registry /bin
```
Comment (jmontleon) [inline contrib/cirrus/setup_environment.sh:39]: Is there a better way to do this?

**Comment by rhatdan** [comment]
LGTM

File: pkg/api/handlers/libpod/manifests.go [hunk 3]
```diff
@@ -162,13 +162,36 @@ func ManifestAdd(w http.ResponseWriter, r *http.Request) {
...
+		TLSVerify bool `schema:"tlsVerify"`
 	}{}
 	if err := json.NewDecoder(r.Body).Decode(&query); err != nil {
 		utils.Error(w, http.StatusInternalServerError, errors.Wrap(err, "Decode()"))
 		return
 	}
 
+	authconf, authfile, err := auth.GetCredentials(r)
+	if err != nil {
+		utils.Error(w, http.StatusBadRequest, errors.Wrapf(err, "failed to parse registry header for %s", r.URL.String()))
```
Comment (mtrmac) [inline pkg/api/handlers/libpod/manifests.go:175]: `GetCredentials` already adds a pretty much the same “failed to parse registry header” wrapper, so this `Wrapf` seems unnecessary.

(I realize there is exactly one instance of doing that, in this file. It probably slipped through during a cleanup.)

File: pkg/api/handlers/libpod/manifests.go [hunk 4]
```diff
@@ -350,6 +373,25 @@ func ManifestModify(w http.ResponseWriter, r *http.Request) {
...
```
Comment (mtrmac) [inline pkg/api/handlers/libpod/manifests.go:378]: This `Wrapf` seems unnecessary, see elsewhere.

File: pkg/api/handlers/libpod/manifests.go [hunk 5]
```diff
@@ -162,13 +162,36 @@ func ManifestAdd(w http.ResponseWriter, r *http.Request) {
...
+	defer auth.RemoveAuthfile(authfile)
+	var username, password string
+	if authconf != nil {
+		username = authconf.Username
+		password = authconf.Password
+	}
+	query.ManifestAddOptions.Authfile = authfile
+	query.ManifestAddOptions.Username = username
+	query.ManifestAddOptions.Password = password
+	query.ManifestAddOptions.All = query.All
```
Comment (mtrmac) [inline pkg/api/handlers/libpod/manifests.go:187]: This seems unrelated to the description of the PR. Is it intentional?

File: pkg/api/handlers/libpod/manifests.go [hunk 6]
```diff
@@ -350,6 +373,25 @@ func ManifestModify(w http.ResponseWriter, r *http.Request) {
...
+	defer auth.RemoveAuthfile(authfile)
+	var username, password string
+	if authconf != nil {
+		username = authconf.Username
+		password = authconf.Password
+	}
+	body.ManifestAddOptions.Authfile = authfile
+	body.ManifestAddOptions.Username = username
+	body.ManifestAddOptions.Password = password
+	body.ManifestAddOptions.All = body.All
```
Comment (mtrmac) [inline pkg/api/handlers/libpod/manifests.go:390]: This seems unrelated to the description of the PR. Is it intentional?

**Review by mtrmac — COMMENTED** [review]
(Disclaimer: I’m not very familiar with the HTTP code)

File: pkg/api/handlers/libpod/manifests.go (hunk 5 above)
Comment (jmontleon) [reply pkg/api/handlers/libpod/manifests.go:187]: @mtrmac Sorry for the stupid question, did you mean specifically the last line here `	query.ManifestAddOptions.All = query.All` or all of it
```
	query.ManifestAddOptions.Authfile = authfile
	query.ManifestAddOptions.Username = username
	query.ManifestAddOptions.Password = password
	query.ManifestAddOptions.All = query.All
```

Comment (mtrmac) [reply pkg/api/handlers/libpod/manifests.go]: I'm sorry, I meant just the `.All` line.

(I’m not saying it’s incorrect, or unwanted, I haven’t checked; it just seems unrelated and I can‘t see any mention of that part.)

Comment (jmontleon) [reply pkg/api/handlers/libpod/manifests.go]: It seems unnecessary. I removed it and tests are passing. I was working through getting credentials passed through properly and wasn't sure based on looking at the ImagePushOptions for Push whether this should also be added.
